core/lib/spree/core/permalinks.rb

```diff
@@ -47,18 +47,14 @@ def generate_permalink
         "#{self.class.permalink_prefix}#{Array.new(self.class.permalink_length){rand(9)}.join}"
       end
 
-      def save_permalink(permalink_value=self.to_param)
+      def save_unique_permalink(permalink_value=self.to_param)
         self.with_lock do
           permalink_value ||= generate_permalink
-
           field = self.class.permalink_field
             # Do other links exist with this permalink?
             other = self.class.where("#{self.class.table_name}.#{field} LIKE ?", "#{permalink_value}%")
             if other.any?
-              # Find the existing permalink with the highest number, and increment that number.
-              # (If none of the existing permalinks have a number, this will evaluate to 1.)
-              number = other.map { |o| o.send(field)[/-(\d+)$/, 1].to_i }.max + 1
-              permalink_value += "-#{number.to_s}"
+              save_unique_permalink(permalink_value)
```

Comment:
two concerns here:
1) if you pass through the previously generated permalink that has a dupe, it doesn't look like a new one would ever get generated
2) I'm not so sure about calling this recursively within a lock and what that will do for the locking strategy. You might need to release the lock before calling this method again.